### PR TITLE
general: applet multiprocess

### PR DESCRIPTION
**Pull request by liamwhite**
This pull request has been a long time in the making. But over the years, blocker after blocker has been removed, and now yuzu is able to run multiple guest processes at the same time! (Actually, it was able to back in December, but we wanted to wait until we had something to show off first before announcing it.)

This PR consists of two broad parts, which are mostly independent:
1. Presentation rewrite. This was the motivation for #12654 (and later #12685), and was needed to get support for multi-layer composition working.
2. Applet manager rewrite. When the SMMU branch was in development, it was clear that in order to have multiple applets working at the same time, we would need to have support for separate instances of the applet state. Fortunately, some recent factors made this pretty straightforward to do as we can now access the applet structure through its applet resource user ID, instead of just assuming a single global applet.

I would like to thank @FernandoS27 and @german77 for their work on GPU and input tasks throughout the development of this project--it would not have been possible without them.

To demonstrate what we can now do with this branch, here are some videos that show things which previously were not working very well with the native applets:
1. Software keyboard: support for inputting additional languages and customized prompts

https://github.com/yuzu-emu/yuzu/assets/9658600/11987ec9-d0f4-4545-a8ef-15db179d8257

2. Inline software keyboard: mostly had the same issues as the normal software keyboard, shown here using a homebrew
keyboard-inline

3. Mii editor: now usable from in game

https://github.com/yuzu-emu/yuzu/assets/9658600/213a397c-1c32-46e2-8425-3b277a2cabb4

4. Offline web applet: usable from in game, but currently has some rendering errors

https://github.com/yuzu-emu/yuzu/assets/9658600/cc6a80e0-cc72-497a-b5a7-2db11938cbe1

5. AllForegroundInitiallyHidden offline web applet: usable from controller minus menu in Super Mario 3D All-Stars collection.

https://github.com/yuzu-emu/yuzu/assets/9658600/9118317c-98e5-4fa1-ad88-728849d68243

And a demonstration of what will be possible in the future:

https://github.com/yuzu-emu/yuzu/assets/9658600/04150c10-77ae-4395-a870-497f960ed1b9

Known issues that should be fixed before tagging or merging:
- [x] Allow switching between native and emulated versions of applets. This is especially important for the guest controller applet, which is not very useful for us. (Done)
- [x] Rendering errors in web applet. Need some time to properly investigate why this happens. (Fixed, issue with software blit engine)
- [x] Incorrect blending with inline software keyboard. When rendered using the normal strategy (not indirect layer strategy), the top half of the layer is transparent, as expected. However, if we try to naively implement alpha blending, we break games and applets which expect to have their alpha channels discarded. Figure out how this works. (Done)
- [x] FSR's EASU pass does not support alpha transparency at all, which means we would have to composite first and then upscale, which might lead to an overall loss in quality in some cases. (Workaround implemented)
- [x] Touch breaks when returning from a library applet to its caller applet. This should get fixed when the touch rewrite is published. (Done, touch rewrite posted)
- [x] Issues with NCE on Android (Fixed)

Tasks to do later:
- [ ] Implement indirect layer support for inline software keyboard

### CODE REVIEW

**Comment by sakuka47** [comment]
Mii creation is fine under mk8d, under Miitopia it's more difficult, FPS drops and when creating the name, if I press + to validate, yuzu crashes, if I validate the name with the "validated" option it works. 

I haven't tested the applet.web under SMO and ssbu.

**Comment by liamwhite** [comment]
@sakuka47 I pushed a fix for the issues in Miitopia

**Comment by sakuka47** [comment]
Hi, 

I tested the new version, I saw that using the experimental driver for controller switch pro makes errors "input joycon" for mk8d, the controller disconnects and the applet has trouble loading.

For Miitopia, the applet loads but takes a few seconds.

If the Experimental Driver is deactivated, all works perfectly and without slowdown for the 2 games mentioned.

As I don't have Mario golf, I wasn't able to test it.

I saw that there was a new "applet" tab. Is there anything to test between the custom applet and the real one?

**Comment by liamwhite** [comment]
> I saw that there was a new "applet" tab. Is there anything to test between the custom applet and the real one?

Custom implementation is the behavior on current master. Real applet is what is enabled by this PR. It can be advantageous to use the custom implementation in some cases because it provides access to features like HLE controller configuration.

**Comment by sakuka47** [comment]
Hello,

Using Qualcomm or turnip driver, Mii validation makes game crash under yuzu android

Config:

0ne plus 11 5G 16Go , Snapdragon 8 gen 2

https://github.com/yuzu-emu/yuzu/assets/93493413/30304cf7-9620-460a-a08b-ebc081c4f9c9

**Comment by liamwhite** [comment]
@sakuka47 should be fixed now

**Comment by sakuka47** [comment]
@liamwhite ,
I've just tested it and it works perfectly. If you want more tests, don't hesitate. 

Thanks you 😁

**Comment by gidoly** [comment]
games like Ni no kuni white witch and prinny can i become hero 1,2  is not playable due to game uses web applet to save progress

web_fix_EtXf0rz5nA

File: src/common/settings.h [hunk 1]
```diff
@@ -133,6 +133,38 @@ struct TouchFromButtonMap {
...
+    Setting<AppletMode> controller_applet_mode{linkage, AppletMode::HLE, "controller_applet_mode",
+                                               Category::LibraryApplet};
+    Setting<AppletMode> data_erase_applet_mode{linkage, AppletMode::HLE, "data_erase_applet_mode",
+                                               Category::LibraryApplet};
+    Setting<AppletMode> error_applet_mode{linkage, AppletMode::HLE, "error_applet_mode",
+                                          Category::LibraryApplet};
+    Setting<AppletMode> net_connect_applet_mode{linkage, AppletMode::HLE, "net_connect_applet_mode",
+                                                Category::LibraryApplet};
+    Setting<AppletMode> player_select_applet_mode{
+        linkage, AppletMode::HLE, "player_select_applet_mode", Category::LibraryApplet};
```
Comment (FernandoS27) [inline src/common/settings.h:148]: this one should be LLE. The LLE version is more straight forward.

Comment (liamwhite) [reply src/common/settings.h:148]: It's currently got some modes in it that breaks games like Big Brain Academy.

**Comment by brujo5** [comment]
Ni no Kuni: Wrath of the White it does not start in this pr build on android


https://github.com/yuzu-emu/yuzu/assets/4172898/edc39f89-5d7c-4180-a486-136d2d9def7d

[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/14133763/yuzu_log.txt)

**Comment by liamwhite** [comment]
@brujo5 your save data import has bad permissions which is causing fs to return errors to the game
```
[  54.324825] Common.Filesystem <Error> common/fs/fs.cpp:CreateDirs:220: Failed to create the directories at path=/storage/emulated/0/Android/data/org.yuzu.yuzu_emu/files/nand/user/save/0000000000000000/B893A8DDAFCEBFA454636ED93AF887E1/0100E5600D446000, ec_message=Permission denied
```

**Comment by brujo5** [comment]
> @brujo5 your save data import has bad permissions which is causing fs to return errors to the game
> 
> ```
> [  54.324825] Common.Filesystem <Error> common/fs/fs.cpp:CreateDirs:220: Failed to create the directories at path=/storage/emulated/0/Android/data/org.yuzu.yuzu_emu/files/nand/user/save/0000000000000000/B893A8DDAFCEBFA454636ED93AF887E1/0100E5600D446000, ec_message=Permission denied
> ```

It's true, it's strange that it has been corrupted since I used a backup with today's build v224

Screenshot_20240201-202909_yuzu Debug Release

Game still require save data for bypass loop in menu

**Comment by brujo5** [comment]
The Lara Croft Collection It was said that this game was fixed with this PR.  but it only fixed the shader compilation that restarted but the game still doesn't start and crashes


https://github.com/yuzu-emu/yuzu/assets/4172898/d0c15aea-6483-481e-adc7-0de7ddd73ff9



[yuzu_log.txt.old.txt](https://github.com/yuzu-emu/yuzu/files/14134651/yuzu_log.txt.old.txt)

**Comment by liamwhite** [comment]
Is that also the case on desktop?

**Comment by EmulationChannel** [comment]
testing in Assassin's Creed® Ⅲ Remastered [ 126.595199] Core.ARM <Critical> core\arm\dynarmic\arm_dynarmic_64.cpp:ExceptionRaised:138: Cannot execute instruction at unmapped address 0xff922b7df56d6106
[ 126.595203] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:16: Backtrace, sp=00000010C3960C60, pc=FF922B7DF56D610A
[ 126.595205] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:18: Module Name         Address             Original Address    Offset              Symbol
[ 126.595206] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:19: 
[ 126.651847] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      FF922BC78283010A    FF922B7DF56D610A    FF922B568283010A    
[ 126.651852] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      0000007100001930    0000002772EA7930    0000000000001930    nns::input::ControllerManager::Update()
[ 126.651853] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      0000007100024298    0000002772ECA298    0000000000024298    SimpleGamePad::Update()
[ 126.651854] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      0000007100024210    0000002772ECA210    0000000000024210    SimpleGamePad::EnterGamePadDetectionThread(void*)
[ 126.651855] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: nnSdk               00000071000AF868    0000000083053868    00000000000AF868    
[ 126.651856] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: nnSdk               00000071000B3BC8    0000000083057BC8    00000000000B3BC8    
[ 126.651857] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: unknown             0000007100000000    0000000000000000    0000000000000000    
[ 126.651915] Core.ARM <Critical> core\arm\dynarmic\arm_dynarmic_64.cpp:ExceptionRaised:138: Cannot execute instruction at unmapped address 0x00000000000000
[ 126.651917] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:16: Backtrace, sp=00000010840008C0, pc=0000000000000004
[ 126.651918] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:18: Module Name         Address             Original Address    Offset              Symbol
[ 126.651919] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:19: 
[ 126.705850] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: unknown             0000007100000004    0000000000000004    0000000000000004    
[ 126.705855] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: nnSdk               00000071003AF2DC    00000000833532DC    00000000003AF2DC    
[ 126.705856] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      0000007100029E68    0000002772ECFE68    0000000000029E68    nn::gfx::detail::QueueImpl<nn::gfx::ApiVariation<nn::gfx::ApiType<4>, nn::gfx::ApiVersion<8>>>::Present(nn::gfx::detail::SwapChainImpl<nn::gfx::ApiVariation<nn::gfx::ApiType<4>, nn::gfx::ApiVersion<8>>>*, int)
[ 126.705857] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      0000007100017D3C    0000002772EBDD3C    0000000000017D3C    GraphicsSystem::Draw()
[ 126.705858] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: Launcher_f.nrs      0000007100018DCC    0000002772EBEDCC    0000000000018DCC    LauncherMain(int, void*, unsigned long, unsigned long, char const*, void*, int&)
[ 126.705859] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: bootloader.nss      00000071000043FC    00000000800083FC    00000000000043FC    nnMain
[ 126.705860] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: nnSdk               000000710008492C    000000008302892C    000000000008492C    nn::init::Start(unsigned long, unsigned long, void (*)(), void (*)())
[ 126.705861] Core.ARM <Error> core\arm\arm_interface.cpp:LogBacktrace:23: nnrtld              00000071000000AC    00000000800000AC    00000000000000AC

**Comment by liamwhite** [comment]
@EmulationChannel is this a new regression caused by this branch, or did this problem already exist?

**Comment by EmulationChannel** [comment]
@liamwhite yes it was caused by this PR but the problem only occurs in this build of this PR test that I carried out in this specific game
​

**Comment by liamwhite** [comment]
I tried ACIII, it does not go in game but it was not regressed by this PR. It still wasn't working with mainline builds from several weeks ago.

**Comment by EmulationChannel** [comment]
@liamwhite Test Assassin's Creed® Ⅲ Remastered but the Liberation HD Remastered version can reach the game with this PR by activating auto draft in the emulator debug settings

**Comment by liamwhite** [comment]
what

**Comment by EmulationChannel** [comment]
@liamwhite I was able to run DLC Liberation game Assassin's Creed® Ⅲ Remastered with your PR, thanks for the correction
​

**Comment by EmulationChannel** [comment]
I meant stub services that ignore some things for the game to run yuzu
​

**Comment by brujo5** [comment]
> The Lara Croft Collection It was said that this game was fixed with this PR. but it only fixed the shader compilation that restarted but the game still doesn't start and crashes
> 
>  VID_20240201_205136_572.mp4 
> [yuzu_log.txt.old.txt](https://github.com/yuzu-emu/yuzu/files/14134651/yuzu_log.txt.old.txt)

Log from Build 226 show more things
Game boot for few seconds then FPS drop to 0


https://github.com/yuzu-emu/yuzu/assets/4172898/9f7796a7-3210-4f75-bbfc-8a88ee654d5f



[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/14146275/yuzu_log.txt)

**Comment by german77** [comment]
Crash due unimplemented `ISelfController::SetMediaPlaybackState`

**Comment by brujo5** [comment]
> @liamwhite I was able to run DLC Liberation game Assassin's Creed® Ⅲ Remastered with your PR, thanks for the correction​

Assassin's creed 3 remastered work great on build 227
Main game + DLC work fine

Screenshot_20240202-202909_yuzu
Screenshot_20240202-205247_yuzu
Screenshot_20240202-205157_yuzu

Screenshot_20240202-203446_yuzu


Screenshot_20240202-204119_yuzu
Screenshot_20240202-202837_yuzu
Screenshot_20240202-204243_yuzu
Screenshot_20240202-204220_yuzu

**Comment by brujo5** [comment]
I tried latest artifac build with SetMediaPlaybackState for self controller but now lara croft collection crashes in shader compilation

[yuzu_log.txt.old.txt](https://github.com/yuzu-emu/yuzu/files/14150452/yuzu_log.txt.old.txt)

Log show new error

**Comment by liamwhite** [comment]
Your firmware install is bad, so it doesn't load the applets at all. Does not seem relevant to this PR.

**Comment by brujo5** [comment]
> Your firmware install is bad, so it doesn't load the applets at all. Does not seem relevant to this PR.

Sorry for the inconvenience but it seems that when loading the official build backups the imported data is somehow corrupted and I have to do it manually.

The game starts and everything but when you select any of the games when you get to the menu and start the game, the game becomes extremely slow and ends up crashing

Screenshot_20240203-000118_yuzu Debug Release
Screenshot_20240202-235717_yuzu Debug Release
Screenshot_20240202-235645_yuzu Debug Release

[yuzu_log.txt.old.txt](https://github.com/yuzu-emu/yuzu/files/14150527/yuzu_log.txt.old.txt)

**Comment by legend800** [comment]
Please stop spamming this thread with unrelated comments and posting numerous pictures.

**Comment by EmulationChannel** [comment]
@liamwhite not fixed new commit stub controller ISelfController::SetMediaPlaybackState
https://cdn.discordapp.com/attachments/1040390872543539200/1203285830840487946/Captura_de_tela_2024-02-03_0727455.png?ex=65d08a42&is=65be1542&hm=934bfb308ae8a6f267ee4a0e73156f165edf8ed9d958d414b574db3549a85c50&

**Comment by german77** [comment]
@EmulationChannel your screenshot indicates we need to implement more things. Please provide a new log

**Comment by EmulationChannel** [comment]
@german77 Oky new logs @liamwhite 
[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/14151721/yuzu_log.txt)

**Comment by brujo5** [comment]
> @liamwhite not fixed new commit stub controller ISelfController::SetMediaPlaybackState https://cdn.discordapp.com/attachments/1040390872543539200/1203285830840487946/Captura_de_tela_2024-02-03_0727455.png?ex=65d08a42&is=65be1542&hm=934bfb308ae8a6f267ee4a0e73156f165edf8ed9d958d414b574db3549a85c50&

If you get the Ac3 game with a black screen after the ubisoft logo, try updating the firmware to the latest version, this is how other users and I made it work on Android in the latest version of Yuzu.

Screenshot_20240203-080635_Nekogram

**Comment by EmulationChannel** [comment]
new log testing 
[yuzu_log.txt.old.txt](https://github.com/yuzu-emu/yuzu/files/14164582/yuzu_log.txt.old.txt)

**Comment by liamwhite** [comment]
I strongly suspect the ACIII issue is a race condition and is unlikely to be fixable on desktop, but will work with NCE.

**Comment by EmulationChannel** [comment]
Version<8>>>::Present(nn::gfx::detail::SwapChainImpl<nn::gfx::ApiVariation<nn::gfx::ApiType<4>,

**Comment by Alkaid-Benetnash** [comment]
When testing this PR, I observed the following error when returning to game "Ring Fit Adventure" from the AllForegroundInitiallyHidden offline web applet:
```
[ 376.098848] Service.AM <Info> core/hle/service/am/library_applet_self_accessor.cpp:PushInteractiveOutData:138: called
[ 376.109458] Service.AM <Info> core/hle/service/am/library_applet_self_accessor.cpp:PopInteractiveInData:122: called
[ 376.166086] Service.AM <Warning> core/hle/service/am/application_functions.cpp:SetTerminateResult:243: (STUBBED) called, result=0xFFFFFFFF
[ 376.166367] Debug.Emulated <Critical> core/hle/kernel/svc/svc_exception.cpp:Break:51: Userspace PANIC! info1=0x00000010C04F6D30, info2=0x0000000000000004
[ 376.166370] Debug.Emulated <Critical> core/hle/kernel/svc/svc_exception.cpp:operator():33: debug_buffer_err_code=FFFFFFFF
[ 376.166380] Debug.Emulated <Critical> core/hle/kernel/svc/svc_exception.cpp:Break:99: Emulated program broke execution! reason=0x0000000000000000, info1=0x00000010C04F6D30, info2=0x0000000000000004
[ 376.166381] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:16: Backtrace, sp=00000010C04F6B80, pc=0000000082C0B4D4
[ 376.166385] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:17: Module Name         Address             Original Address    Offset              Symbol
[ 376.166385] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:19: 
[ 376.167425] Service.AM <Warning> core/hle/service/am/display_controller.cpp:AcquireLastApplicationCaptureSharedBuffer:79: (STUBBED) called
[ 376.248751] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               00000071000FC4D4    0000000082C0B4D4    00000000000FC4D4    
[ 376.248755] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               0000007100056518    0000000082B65518    0000000000056518    nn::diag::detail::Abort(nn::Result const*)
[ 376.248757] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               0000007100054E10    0000000082B63E10    0000000000054E10    
[ 376.248757] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               0000007100054ED8    0000000082B63ED8    0000000000054ED8    
[ 376.248758] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               000000710006596C    0000000082B7496C    000000000006596C    
[ 376.248759] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: Billy.nss           0000007100861F48    0000000080865F48    0000000000861F48    
[ 376.248760] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: Billy.nss           000000710083B354    000000008083F354    000000000083B354    
[ 376.248761] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: Billy.nss           000000710083A428    000000008083E428    000000000083A428    
[ 376.248762] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: Billy.nss           0000007100F360E0    0000000080F3A0E0    0000000000F360E0    
[ 376.248763] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: Billy.nss           0000007100F353D8    0000000080F393D8    0000000000F353D8    
[ 376.248765] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               00000071000CFB48    0000000082BDEB48    00000000000CFB48    
[ 376.248766] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: nnSdk               00000071000D4314    0000000082BE3314    00000000000D4314    
[ 376.248767] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:22: unknown             0000007100000000    0000000000000000    0000000000000000    
[ 376.289171] Service.AM <Warning> core/hle/service/am/display_controller.cpp:ReleaseLastApplicationCaptureSharedBuffer:91: (STUBBED) called
[ 376.289604] Service.AM <Info> core/hle/service/am/window_controller.cpp:SetAppletWindowVisibility:56: called
[ 376.289842] Service.AM <Warning> core/hle/service/am/window_controller.cpp:SetAppletGpuTimeSlice:80: (STUBBED) called, time_slice=100000
[ 376.296757] Service.AM <Warning> core/hle/service/am/display_controller.cpp:TakeScreenShotOfOwnLayer:72: (STUBBED) called
[ 376.350629] Debug <Critical> ./video_core/texture_cache/texture_cache.h:operator():2031: Assertion Failed!
Unregistering unregistered image in page=0x502200000
[ 376.350632] Debug <Critical> ./video_core/texture_cache/texture_cache.h:operator():2031: Assertion Failed!
Unregistering unregistered image in page=0x502300000
[ 376.350633] Debug <Critical> ./video_core/texture_cache/texture_cache.h:operator():2031: Assertion Failed!
```

Full log is here [yuzu_ring_fit_leg_strap_setting_stuck.txt](https://github.com/yuzu-emu/yuzu/files/14207604/yuzu_ring_fit_leg_strap_setting_stuck.txt)

To reproduce this: go to "My Menu" -> Settings -> Leg Strap Settings.

According to the behavior observed in my local qt websession PoC, after clicking this leg strap settings button in the web applet, the game (not the applet) is supposed to be wake up and run a joycon motion sensors calibration process.

Not sure what goes wrong here.

**Comment by german77** [comment]
motion calibration isn't supported at the moment this is not an issue with the applet

**Comment by EmulationChannel** [comment]
@liamwhite Thanks fixed Assassin's Creed® Ⅲ Remastered and DLCS Game https://files.fm/u/9tvy9253zs

**Review by german77 — APPROVED** [review]

**Comment by benderscruffy** [comment]
Funimation [01008e10130f8000][v0] fails to boot now with this pr

**Comment by liamwhite** [comment]
@benderscruffy you need to provide a log

**Pull request merged**